Initial changes for Open XL z/OS support

Deigue opened the pull request:
Includes all the initial work done a while ago made in order to support compiling with Open XL in context of z/OS.

(This is one part of the multiple changes added for supporting Open XL compilation on OMR)

keithc-ca commented on thread/common/thrprof.c:
```diff
@@ -470,7 +470,11 @@ omrthread_get_handle(omrthread_t thread)
 		In order to work around the compiler error, we have to reach inside
 		the structure do the dirty work. The handle may not even be correct! */
 	uintptr_t *tempHandle;
-	tempHandle = (uintptr_t *)&(thread->handle.__[0]);
+	#if defined(__GNUC__)
+		tempHandle = (uintptr_t *)&(thread->handle.__);
+	#else
+		tempHandle = (uintptr_t *)&(thread->handle.__[0]);
+	#endif
```
This should use the same pattern as in https://github.com/eclipse/omr/pull/7323.

Deigue replied:
So I am seeing https://github.com/Deigue/omr/blob/4401fae20ac9db1255fafb68333e322297d43620/include_core/thrtypes.h#L57 , `handle` is of type `OSTHREAD` 

```
OSTHREAD tempHandle = thread->handle;
return  (uint_ptr)*  <cast> tempHandle
```
But what would the cast look like? Where can I find `OSTHREAD` to see what exactly __ is pulling out and how to cast it properly? "handle" also seemed like too broad a search/grep for `/usr/include`

keithc-ca replied:
Something like this is what I meant:
```c
    OSTHREAD tempHandle = thread->handle;
    return (uintptr_t)*(unsigned long long *)&tempHandle;
```
For z/OS, `OSTHREAD` is defined in `include_core/unix/thrdsup.h`:
```c
typedef pthread_t OSTHREAD;
```

Deigue replied:
Thanks, confirmed it works and everything compiles, and changed accordingly.

babsingh commented:
jenkins build all

Deigue commented:
java21 passed: https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK21_s390x_zos_Personal/569/parameters/
However, will be making more additional changes, so will need reverify here after have a more final draft.

Deigue commented:
[j21 jenkins verify](https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK21_s390x_zos_Personal/648/) Verified new changes compiling with XLC as well, after checking locally working with Open XL. Removing draft and changing to ready for review ...

keithc-ca commented on include_core/omrsig.h:
```diff
@@ -42,7 +42,9 @@ typedef __sighandler_t sighandler_t;
 typedef void (*sighandler_t)(int sig);
 #elif defined(J9ZOS390) || defined(AIXPPC)
 typedef void (*sighandler_t)(int sig);
+#if !defined(__GNUC__)
 #define __THROW
+#endif
```
This is within an `extern "C"` block, so none of the uses of `__THROW` below seem appropriate.
The macro definition, and its uses here and in `omrsig.cpp` should just be removed.

Deigue replied:
It is extern "C" only if `__cplusplus` is defined right?

And if that is the case, the it is safe to remove the define instances below entirely?
L35: `#define __THROW`
Snippet L35:37 
```
#if defined(OSX)
#define __THROW
#endif /* defined(OSX) */
```

And what does that mean for these below where __THROW is being used, does the "__THROW" part need to be removed specifically from every line? https://github.com/eclipse/omr/blob/c256823419b4579cf64f94072337fa108f1761f9/include_core/omrsig.h#L104-L115

keithc-ca replied:
The definitions of `__THROW` should be removed as well as all uses.

Deigue replied:
I get these errors when removing the specifications:
```
/omr/omrsigcompat/omrsig.cpp:241:1: error: 'signal' is missing exception specification 'throw()'
  241 | signal(int signum, sighandler_t handler)
      | ^
      |                                          throw()
```

This being shown for all 5 functions defined in omrsig.cpp with `__THROW`

keithc-ca replied:
I don't see the suggested change to `omrsig.cpp`. I repeat my objection to declaring `throw()` for C function declarations.

keithc-ca commented on port/common/omrgetsysname.h:
```diff
@@ -25,6 +25,7 @@
 
 #include "omrcomp.h"
 
+struct OMRPortLibrary;
```
A blank line after this one would be welcome.

Deigue replied:
Done

keithc-ca commented on omrsigcompat/makefile:
```diff
@@ -24,7 +24,7 @@ top_srcdir := ..
 
 MODULE_NAME := omrsig
-ARTIFACT_TYPE := c_shared
+ARTIFACT_TYPE := cxx_shared
```
This has nothing to do with OpenXL; it's just an error that could be corrected separately.

Deigue replied:
removed from this PR.

keithc-ca commented on port/unix/omrshmem.c:
```diff
@@ -1663,7 +1663,7 @@ omrshmem_openSharedMemory (OMRPortLibrary *portLibrary, intptr_t fd, const char
 			}
 		} else {
-#if defined(__GNUC__) || defined(AIXPPC)
+#if (defined(__GNUC__) || defined(AIXPPC)) && !defined(J9ZOS390)
```
Suggest reordering the `#if` branches so it's clear which one is intended for z/OS, or `#define KEY` appropriately for each platform/compiler and then use `KEY` instead of `key`, `_key` or `__key`.

Deigue replied:
can you detail that a bit, looking at the logic again, since I have excluded via `!defined(J9ZOS390)` , it should mean that z/OS does not use any of the child `if` branches underlying within the parent `if` clause right?

keithc-ca replied:
I see now that none of the branches would apply for z/OS: is that the right behavior?
If so, I suggest we reorder the code to put the z/OS path first, and this in the `#else` block:
```c
#if defined(OMRZOS390)
    // existing lines 1686-1709
#elif defined(__GNUC__) || defined(AIXPPC)
    // existing lines 1667-1683
#endif /* defined(OMRZOS390) */
```

Deigue replied:
yup, and that does look cleaner.
But what is the difference (if any) between `OMRZOS390` and `J9ZOS390` ... because if it seems there is diff, then I will still need it as follows:
```
#if defined(OMRZOS390)
    // existing lines 1686-1709
#elif (defined(__GNUC__) || defined(AIXPPC)) && !defined(J9ZOS390)
    // existing lines 1667-1683
#endif /* defined(OMRZOS390) */
```

keithc-ca replied:
I think `OMRZOS390` is the preferred flag to test in new code, `J9ZOS390` is a name that originated in OpenJ9. They're both still used and I think they're equivalent, but OMR is supposed to be a base for OpenJ9, not dependent on anything from OpenJ9 (including naming conventions).

Deigue replied:
Seems the build is failing after reorganizing this and `omrshsem_deprecated.c` 
```
cd .../omr/build/port && /xlc210/usr/lpp/IBM/cnw/v2r1/openxl/bin/ibm-clang64  -D__STDC_LIMIT_MACROS -D_ALL_SOURCE -D_OPEN_THREADS=3 -D_XOPEN_SOURCE=600 -DJ9ZOS390 -DJ9ZOS39064 -DLONGLONG -DOMR_EBCDIC -DOMRPORT_LIBRARY_DEFINE -DSUPPORTS_THREAD_LOCAL -DZOS -I...(all includes here)  -fstrict-aliasing -mzos-target=ZOSV2R4 -m64 -march=arch10   -fvisibility=default -o CMakeFiles/omrport_obj.dir/unix/omrshmem.c.o   -c /.../omr/port/unix/omrshmem.c  
```

The full command follows the above format, and `J9JZ0S390` is defined, but looks like `OMRZOS390` is not ... so I might need to switch it to `J9JZOS390`. With that, I think the reorganizing should work fine.

keithc-ca replied:
I don't see `OMRZOS390` defined anywhere (despite tests in this file and in `omrsharedhelper.h`).
Unless someone can point to where `OMRZOS390` is defined, this should use `J9ZOS390`.

Deigue replied:
Agreed, thats what I was mentioning in my prior comment. As such, in the current state of the PR I am still utilizing `J9ZOS390` in this file.

keithc-ca replied:
The reordering I [suggested](https://github.com/eclipse-omr/omr/pull/7320/files#r1646575284) seems to have disappeared.

keithc-ca replied:
And the z/OS branch still uses `OMRZOS390` instead of `J9ZOS390`.

keithc-ca commented on port/unix/omrshsem_deprecated.c:
```diff
@@ -1060,7 +1060,7 @@ omrshsem_openSemaphore(struct OMRPortLibrary *portLibrary, intptr_t fd, char *ba
 			}
 		} else {
-#if defined(__GNUC__) || defined(AIXPPC) || defined(OMRZTPF)
+#if (defined(__GNUC__) || defined(AIXPPC) || defined(J9ZTPF)) && !defined(J9ZOS390)
```
Same comment regarding `KEY` here.

keithc-ca replied:
Please add the updated condition to line 1084:
```c
#endif /* (defined(__GNUC__) || defined(AIXPPC) || defined(J9ZTPF)) && !defined(J9ZOS390) */
```

keithc-ca commented on port/zos390/omrgetthent.h:
```diff
@@ -28,7 +28,7 @@
 #define PGTHA_FLAG_PROCESS_DATA 0x80
 
-#pragma pack(packed)
+#pragma pack(push, 1)
```
Does this mean the same for both xlc and OpenXL?

Deigue replied:
yes they should be functionally the same.
https://www.ibm.com/docs/en/zos/2.4.0?topic=descriptions-pragma-pack
also have confirmed internally with Sean from the open-xl/cpp team

Deigue replied:
Have changed this to `#pragma pack(1)` after having confirmed the packed struct sizes between XLC and Open XL and confirming the functionality is identical with Sean. He suggested to use this notation instead.

keithc-ca commented on port/zos390/omrintrospect.h:
```diff
@@ -34,22 +34,26 @@
 #pragma map(pthread_quiesce, "BPX1PTQ")
 #endif
 
+#if defined(OMR_ENV_DATA64)
+#else
```
Just say `#if !defined(OMR_ENV_DATA64)` and add the missing `#endif` comment.

Deigue replied:
looks like I have your feedback in the other PR: https://github.com/eclipse/omr/pull/7321
Might just remove this file from this PR, as that one seems to handle more updated/latest code for `omrintrospect.h`

keithc-ca commented on port/zos390/omrsimap.h:
```diff
@@ -48,7 +48,7 @@
  * The RSM Internal Table (RIT) reached from the `ritp' address field in PVT. The
  * field of interest here is the total amount of online storage.
  */
-typedef _Packed struct J9RIT {
+typedef struct __attribute__((packed)) J9RIT {
```
Again, do xlc and OpenXL attach the same meaning to this?

Deigue replied:
They do have the same meaning as per discussing internally with Sean from the cpp team. I did inquire if we have any published/public documentation for this, but I dont think they have a dedicated page regarding this particular suggestion/tweak.
However digging a bit and searching for it, there is a small mention of it on this page: https://www.ibm.com/docs/en/open-xl-c-cpp-zos/2.1?topic=guide-binary-compatibility

To quote part of our discussions: 

> The problem with _Packed is that it is C only and the usual change to attribute((packed)) for C++ doesn't have the same semantics.
```c
typedef struct S _Packed {
   char c;
   int i;
 } S;
 S s1;
 struct S s2;
 ```
> In this example s1 and s2 have different sizes and types.  The _Packed keyword applies to the declarator (eg. typedef name S) on the struct definition.  The _Packed keyword creates a hidden duplicate version of the struct.
> Avoid the keyword even in XL.

keithc-ca replied:
Builds have been failing with `__attribute__((packed))` present. I don't think that's needed with the addition of the pragmas. I think the pattern we want is:
```c
#pragma pack(1)
typedef struct J9RIT { /* ... */ } J9RIT;
/* etc. */
#pragma pack(pop)
```

Deigue commented:
@keithc-ca This one is also ready. I have added comments for prior questions above. Wanted to make sure this is approved from your perspective before Babneet can kick off the jenkins builds and start the process for merging the changes.

Deigue commented:
As per latest code changes proposed in PR, I have kept it utilizing `J9ZOS390`, as I confirmed this is the way it compiles properly.

For the __THROW related investigation, I have branched off and created an issue to track: https://github.com/eclipse-omr/omr/issues/7643 to address in a seperate PR in the future.

Since the changes here are strictly only for Open XL enablement, want to try to minimalize this PR itself and get it merged soon as it has been opened for a long period.

Deigue commented:
@keithc-ca I have removed the `omrsig.h` changes and `THROW` related changes altogether from this PR.
I have moved those changes over into a new PR here: https://github.com/eclipse-omr/omr/pull/7766

In order to get things moving along, since the remaining changes were good to go, I have rebased against `origin/master` latest ...

I will also rebuild against:
- omr standalone
-  jdk21/openj9 
- java 8 (platforms: mz,xi ?) Feel free to add any suggested platforms that I should be verifying against with test.sanity

---

Once these clear up, I can bring this to ready for review as well, Let me know if I am missing anything from the above checklist here.

Deigue commented:
java 8: vm farm build id 95535, seeing few job errors, but uncertain if related.
omr standalone: failing currently, but still seeing `THROW` related errors, which is odd since those changes were removed. Might be some cleaning up needed when testing, as the changes dont seem to be picked up right.

keithc-ca commented on port/unix/omrshmem.c:
```diff
@@ -1663,7 +1663,7 @@ omrshmem_openSharedMemory (OMRPortLibrary *portLibrary, intptr_t fd, const char
 			}
 		} else {
-#if defined(__GNUC__) || defined(AIXPPC)
+#if (defined(__GNUC__) || defined(AIXPPC)) && !defined(J9ZOS390)
```
Please repeat the (udpated) condition on line 1684:
```c
#endif /* (defined(__GNUC__) || defined(AIXPPC)) && !defined(J9ZOS390) */
```

keithc-ca commented on thread/common/thrprof.c:
```diff
@@ -502,14 +502,8 @@ uintptr_t
-	uintptr_t *tempHandle;
-	tempHandle = (uintptr_t *)&(thread->handle.__[0]);
-	return *tempHandle;
+	OSTHREAD tempHandle = thread->handle;
+	return (uintptr_t)*(unsigned long long *)&tempHandle;
 #else
 	return (uintptr_t)thread->handle;
 #endif
```
While you're here, please fix lines 507, 509:
```c
#else /*  defined(J9ZOS390) */
	return (uintptr_t)thread->handle;
#endif /*  defined(J9ZOS390) */
```

Deigue commented:
I've made the suggested formatting changes.
I also verified that the jdk21 xlc builds are fine with these updates: Build_JDK21_s390x_zos_Personal/2769/
I also tried to check for Java 8 again(vmfarm build_id=97168), but I think there were connectivity issues and some jobs have failed.
Should be ok since 
> The issues in 95535 are unrelated.
But was trying to check again to make sure.

babsingh commented:
jenkins build all

babsingh commented:
@Deigue See the errors in the z/OS PR build.

Deigue commented:
```
11:59:17  FSUM3065 The COMPILE step ended with return code 4.
11:59:17  FSUM3017 Could not compile /u/user1/workspace/Build/port/unix/omrsysinfo.c. Correct the errors and try again.
11:59:17  port/CMakeFiles/omrport_obj.dir/build.make:1281: recipe for target 'port/CMakeFiles/omrport_obj.dir/unix/omrsysinfo.c.o' failed
11:59:17  make[2]: *** [port/CMakeFiles/omrport_obj.dir/unix/omrsysinfo.c.o] Error 3
11:59:17  make[2]: *** Waiting for unfinished jobs....
11:59:17  [ 19%] Building CXX object fvtest/compilertest/CMakeFiles/testcompiler.dir/__/__/compiler/codegen/Analyser.cpp.o
11:59:17  FSUM3065 The COMPILE step ended with return code 4.
11:59:17  FSUM3017 Could not compile /u/user1/workspace/Build/port/zos390/omrsysinfo_helpers.c. Correct the errors and try again.
11:59:17  port/CMakeFiles/omrport_obj.dir/build.make:1305: recipe for target 'port/CMakeFiles/omrport_obj.dir/zos390/omrsysinfo_helpers.c.o' failed
11:59:17  make[2]: *** [port/CMakeFiles/omrport_obj.dir/zos390/omrsysinfo_helpers.c.o] Error 3
11:59:17  CMakeFiles/Makefile2:691: recipe for target 'port/CMakeFiles/omrport_obj.dir/all' failed
11:59:17  make[1]: *** [port/CMakeFiles/omrport_obj.dir/all] Error 2
11:59:17  make[1]: *** Waiting for unfinished jobs....
```
Seeing this from the above, I wonder if its unique to the omr "standalone" configuration that the issue happens ...

babsingh commented:
Return `code 4` in `FSUM3065` usually means the compiler treated the warnings as errors. The below warnings need to be addressed.

```
11:59:16  [ 19%] Building C object port/CMakeFiles/omrport_obj.dir/unix/omrsysinfo.c.o
11:59:17  [ 19%] Building C object port/CMakeFiles/omrport_obj.dir/zos390/omrsysinfo_helpers.c.o
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:51    The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:51    Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:61    The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:61    Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:76    The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:76    Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:94    The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:94    Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:109   The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:109   Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:125   The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:125   Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:143   The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:143   Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:163   The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:163   Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  [ 19%] Linking CXX executable ../../omrutiltest
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:51    The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:51    Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:61    The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:61    Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:76    The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:76    Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:94    The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:94    Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:109   The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:109   Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:125   The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:125   Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:143   The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:143   Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  INFORMATIONAL CCN4108 /u/user1/workspace/Build/port/zos390/omrsimap.h:163   The use of keyword '__attribute__' is non-portable.
11:59:17  WARNING CCN4105 /u/user1/workspace/Build/port/zos390/omrsimap.h:163   Attribute "packed" is not supported for type specifications and is ignored.
11:59:17  FSUM3065 The COMPILE step ended with return code 4.
11:59:17  FSUM3017 Could not compile /u/user1/workspace/Build/port/unix/omrsysinfo.c. Correct the errors and try again.
11:59:17  port/CMakeFiles/omrport_obj.dir/build.make:1281: recipe for target 'port/CMakeFiles/omrport_obj.dir/unix/omrsysinfo.c.o' failed
11:59:17  make[2]: *** [port/CMakeFiles/omrport_obj.dir/unix/omrsysinfo.c.o] Error 3
11:59:17  make[2]: *** Waiting for unfinished jobs....
11:59:17  [ 19%] Building CXX object fvtest/compilertest/CMakeFiles/testcompiler.dir/__/__/compiler/codegen/Analyser.cpp.o
11:59:17  FSUM3065 The COMPILE step ended with return code 4.
11:59:17  FSUM3017 Could not compile /u/user1/workspace/Build/port/zos390/omrsysinfo_helpers.c. Correct the errors and try again.
11:59:17  port/CMakeFiles/omrport_obj.dir/build.make:1305: recipe for target 'port/CMakeFiles/omrport_obj.dir/zos390/omrsysinfo_helpers.c.o' failed
11:59:17  make[2]: *** [port/CMakeFiles/omrport_obj.dir/zos390/omrsysinfo_helpers.c.o] Error 3
11:59:17  CMakeFiles/Makefile2:691: recipe for target 'port/CMakeFiles/omrport_obj.dir/all' failed
11:59:17  make[1]: *** [port/CMakeFiles/omrport_obj.dir/all] Error 2
11:59:17  make[1]: *** Waiting for unfinished jobs....
```

Deigue commented:
I tried a local build with omr on `stlaba0` 

Expand details:
<details>

OMR build config
```
> cmake -Wdev -C../cmake/caches/Travis.cmake -DCMAKE_C_COMPILER=/bin/c89 -DCMAKE_CXX_COMPILER=/bin/xlc -DOMR_DDR=OFF -DOMR_THR_FORK_SUPPORT=0 ..
loading initial cache file ../cmake/caches/Travis.cmake
-- Starting with CMake version 3.5.1
-- Found Git: /rsusr/miniconda/envs/stlaba0/bin/git (found version "2.14.4-12")
-- The CXX compiler identification is zOS
-- The C compiler identification is zOS
-- Check for working CXX compiler: /bin/xlc
-- Check for working CXX compiler: /bin/xlc -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working C compiler: /bin/c89
-- Check for working C compiler: /bin/c89 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- CMAKE_C_COMPILER is /bin/c89
-- NO XLClang match. Using xlc toolchain.
-- OMR: The CPU architecture is s390
-- OMR: The OS is zos
-- OMR: The tool configuration is xlc
-- OMR: The target data size is 64
-- Looking for dladdr
-- Looking for dladdr - not found
-- The ASM compiler identification is zOS
-- Found assembler: /bin/c89
-- Warning: Did not find file Compiler/zOS-ASM
CMake Warning at cmake/modules/OmrAssert.cmake:58 (message):
  Assertion failed: DEFINED OMR_C_WARNINGS_AS_ERROR_FLAG
Call Stack (most recent call first):
  cmake/modules/platform/toolcfg/verify.cmake:22 (omr_assert)
  cmake/modules/OmrPlatform.cmake:45 (include)
  CMakeLists.txt:50 (include)


CMake Warning at cmake/modules/OmrAssert.cmake:58 (message):
  Assertion failed: DEFINED OMR_CXX_WARNINGS_AS_ERROR_FLAG
Call Stack (most recent call first):
  cmake/modules/platform/toolcfg/verify.cmake:23 (omr_assert)
  cmake/modules/OmrPlatform.cmake:45 (include)
  CMakeLists.txt:50 (include)


-- OMR: trace root is '/jit/team/gauravc/repos/xlc/openj9-openjdk-jdk21-zos/omr/build'
-- DEBUG: RUNTIME_OUTPUT_DIR=/jit/team/gauravc/repos/xlc/openj9-openjdk-jdk21-zos/omr/build
-- DEBUG: CFLAGS= "-Wc,xplink" "-Wc,rostring" "-Wc,FLOAT(IEEE,FOLD,AFP)" "-Wc,enum(4)" "-Wa,goff" "-Wc,NOANSIALIAS" "-Wc,TARGET(ZOSV2R3)" -Wc,lp64 "-Wa,SYSPARM(BIT64)" "-Wc,ARCH(10)" "-Wc,TUNE(12)" "-Wl,compat=ZOSV2R3" "-Wc,langlvl(extc99)"
-- DEBUG: EXE LDFLAGS=
-- DEBUG: so LDFLAGS= -Wl,xplink -Wl,dll -Wl,lp64
-- Found Perl: /rsusr/miniconda/envs/stlaba0/bin/perl (found version "5.24.4")
-- Found PythonInterp: /rsusr/miniconda/envs/stlaba0/bin/python (found version "3.7")
-- Creating static library for jitbuilder
-- Adding PRIVATE compiler options to jitbuilder for jitbuilder.
CMake Warning at fvtest/porttest/CMakeLists.txt:221 (message):
  Excluded omrport tests:
  --gtest_filter=-PortFileTest2.file_test20:PortFileTest2.file_test21:PortFileTest2.file_test22:PortFileTest2.file_test23:PortFileTest2.file_test24:PortFileTest2.file_test25:PortFileTest2.file_test27:PortFileStreamTest.omrfilestream_test_text:PortFileStreamTest.omrfilestream_test_omrfile_text_comparison:PortMmapTest.mmap_test5:PortSigTest.sig_test_async_handler:PortSigTest.sig_test_single_async_handler:PortSigTest.sig_test_mix_async_handler:PortSigTest.sig_test_mix_two_async_handlers:PortStrTest.str_test2:PortStrTest.str_convPlatTo8:PortStrTest.str_convU8ToPlat:PortStrTest.str_convU8ToWide:PortStrTest.str_convWideToU8:PortStrTest.str_convU8ToPlat_Null:PortSysinfoTest.sysinfo_test_get_cwd3:PortSockTest.two_socket_datagram_communication


-- Creating static library for testcompiler
-- Adding PRIVATE compiler options to testcompiler for testcompiler.
-- Adding INTERFACE compiler options to testcompiler for testcompiler.
-- Adding PUBLIC compiler options to tril for testcompiler.
-- Configuring done
-- Generating done
-- Build files have been written to: /jit/team/gauravc/repos/xlc/openj9-openjdk-jdk21-zos/omr/build
```
</details>

I converted to ascii the error logs I have, but I was having some strange linker issues locally on `stlaba0` that would stop me from building omr standalone properly. This may be unrelated though as this used to happen even with `master` branch.

[cmakeoutput-ascii.log](https://github.com/user-attachments/files/21374244/cmakeoutput-ascii.log)
[cmakeerror-ascii.log](https://github.com/user-attachments/files/21374245/cmakeerror-ascii.log)

While actually going through the console build output though, it gets well past the compilation of both `omrsysinfo.c` and `omrsysinfo_helpers` , as can be seen, and then hits the problem that I believe is something environmental in nature:

console build.logs

<details>

```
[ 94%] Building CXX object jitbuilder/CMakeFiles/jitbuilder.dir/__/compiler/z/runtime/VirtualGuardRuntime.cpp.o
[ 94%] Building CXX object fvtest/compilertest/CMakeFiles/compilertest.dir/tests/TestDriver.cpp.o
make[2]: Entering directory '/jit/team/gauravc/repos/xlc/openj9-openjdk-jdk21-zos/omr/build'
Scanning dependencies of target tril_dumper
make[2]: Leaving directory '/jit/team/gauravc/repos/xlc/openj9-openjdk-jdk21-zos/omr/build'
make[2]: Entering directory '/jit/team/gauravc/repos/xlc/openj9-openjdk-jdk21-zos/omr/build'
[ 94%] Building CXX object fvtest/tril/tril/CMakeFiles/tril_dumper.dir/compiler.cpp.o
[ 94%] Linking CXX executable ../../../tril_compiler
[ 94%] Linking CXX static library ../libjitbuilder.a
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/compile/ResolvedMethod.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/compile/ResolvedMethod.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRLinkage.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRLinkage.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRTreeEvaluator.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRTreeEvaluator.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRPeephole.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRPeephole.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRSnippet.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRSnippet.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRCodeGenerator.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRCodeGenerator.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRCodeGenPhase.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRCodeGenPhase.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRMemoryReference.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRMemoryReference.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRMachine.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRMachine.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRRegister.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRRegister.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRRegisterDependency.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRRegisterDependency.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRRealRegister.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRRealRegister.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/codegen/OMRInstruction.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/codegen/OMRInstruction.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/env/OMRCPU.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/env/OMRCPU.cpp.o".
ar: FSUM9942 "CMakeFiles/jitbuilder.dir/__/compiler/z/env/OMRDebugEnv.cpp.o" ignored, same basename as "CMakeFiles/jitbuilder.dir/__/compiler/env/OMRDebugEnv.cpp.o".
[ 94%] Building CXX object fvtest/compilertest/CMakeFiles/compilertest.dir/tests/SingleBitContainerTest.cpp.o
�������မ����(�|<������������������������������������������������������������������������>�|����/��������(���?��`_�?%^^����^���?%���(���?��&�^�ꑑ�?_��%������������/��?>��+���|<�����(�(����|�<�+|����+�<������|(���������+������������������<<�<����������������������������������������������������������������������မ����(�|<����������������������������������������������������������������������_/,�&/�/_����<���^^����^���?%���(���?��&�^������?%���(���?��`_�?������������%��+���|<�����(�(����|�<�+|����+�<������|(���������+�����<<�������������<���������������������������������������������������������������������������မ����(�|<������������������������
```

</details>

 @babsingh Any things I need to be looking out for while building standalone omr , @keithc-ca any specifics environmental or setup wise? I havent been able to build xlc based standalone omr for a while on the machine.

But I have a feeling even the problems above aside, what we are seeing on `jenkins build all` is a different issue altogether, since compiling that is going through without issues.

babsingh commented:
jenkins build zos(compile:'VERBOSE=1')

Deigue commented:
The cause of the issue sounds right, it is odd because I did the return code 4 in my local logs of the incomplete build as well, but it didn't seem to halt the build.
I added the adjusted convention suggested by Keith to `omrsimap.h` as well that Open XL recognizes and XLC is still good with, hopefully it should address the problems.
@babsingh Can you issue another build? just pushed a couple minutes ago

babsingh commented:
jenkins build all

Deigue commented:
I ran a Personal https://[hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK21_s390x_zos_Personal/2941/consoleText](https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK21_s390x_zos_Personal/2941/consoleText) jdk21 now that things are looking better on the omr side,

But seeing a failure that I dont think I've seen before/ potentially unrelated:

```
[2025-07-23T18:57:29.842Z] [ 71%] Building CXX object runtime/compiler/CMakeFiles/j9jit.dir/optimizer/J9EstimateCodeSize.cpp.o
[2025-07-23T18:57:29.842Z] cd /jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm/runtime/compiler && /bin/xlc++ -D__STDC_LIMIT_MACROS -D_AE_BIMODAL -D_ALL_SOURCE -D_OPEN_THREADS=3 -D_XOPEN_SOURCE=600 -Dj9jit_EXPORTS -DBITVECTOR_BIT_NUMBERING_MSB -DBITVECTOR_64BIT -DBUILD_Z_RUNTIME_INSTRUMENTATION -DCOMPRESS_AOT_DATA -DIBM_ATOE -DIPv6_FUNCTION_SUPPORT -DJ9_PROJECT_SPECIFIC -DJ9VM_TIERED_CODE_CACHE -DJ9ZOS390 -DJ9ZOS39064 -DLONGLONG -DMAXMOVE -DOPENJ9_BUILD -DSUPPORTS_THREAD_LOCAL -DTR_HOST_S390 -DTR_HOST_64BIT -DTR_TARGET_S390 -DTR_TARGET_64BIT -DZOS "-Wc,xplink" "-Wc,rostring" "-Wc,FLOAT(IEEE,FOLD,AFP)" "-Wc,enum(4)" "-Wa,goff" "-Wc,NOANSIALIAS" "-Wc,TARGET(zOSV2R3)" -Wc,lp64 "-Wa,SYSPARM(BIT64)" -O3 -qstackprotect "-Wc,inline(auto,noreport,600,5000)" "-Wc,list(),offset,gonumber" -qdebug=nohook -qnogonumber -qxplink=noback -+ "-Wc,ARCH(10)" "-Wc,TUNE(12)" "-Wl,compat=zOSV2R3" "-Wc,langlvl(extended)" -qlanglvl=extended0x -qasm -qnortti -DYYLMAX=1000 -Wa,asa -Wc,EXH -qhaltonmsg=CCN6102 -qnocsect -Wc,DLL,EXPORTALL "-Wc,convlit(ISO8859-1),nose,se(/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/omr/cmake/modules/platform/os/../../../../util/a2e/headers)" -I/usr/lpp/hzc/include -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/compiler/z -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/compiler -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/omr/compiler/z -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/omr/compiler -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/omr -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/omr/thread -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/omr/gc/include -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/compiler/.. -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/codert_vm -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/gc_include -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/gc_glue_java -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/jit_vm -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/nls -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/oti -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/include -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/util -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm/runtime -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm/runtime/oti -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm/runtime/omr -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm/runtime/compiler -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/vendor/runtime/compiler -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/omr/include_core -I/usr/include -I/usr/lpp/cbclib/include -I/include_core -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/vendor/runtime/include -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm/runtime/include -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm/runtime/nls -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/omr/util/hashtable/. -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm/runtime/util -I/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/util/. -o CMakeFiles/j9jit.dir/optimizer/J9EstimateCodeSize.cpp.o -c /jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/compiler/optimizer/J9EstimateCodeSize.cpp
[2025-07-23T18:57:31.713Z] "/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/compiler/optimizer/J9EstimateCodeSize.cpp", line 992.10: CCN5246 (S) The "case" statement bypasses the initialization of "const bool enablePeekingForMHInvokes".
[2025-07-23T18:57:31.713Z] CCN0793(I) Compilation failed for file /jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/openj9/runtime/compiler/optimizer/J9EstimateCodeSize.cpp.  Object file not created.
[2025-07-23T18:57:31.713Z] gmake[6]: *** [runtime/compiler/CMakeFiles/j9jit.dir/build.make:1418: runtime/compiler/CMakeFiles/j9jit.dir/optimizer/J9EstimateCodeSize.cpp.o] Error 12
[2025-07-23T18:57:31.713Z] gmake[6]: Leaving directory '/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm'
[2025-07-23T18:57:31.713Z] gmake[5]: *** [CMakeFiles/Makefile2:8974: runtime/compiler/CMakeFiles/j9jit.dir/all] Error 2
[2025-07-23T18:57:31.713Z] gmake[5]: Leaving directory '/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm'
[2025-07-23T18:57:31.713Z] gmake[4]: *** [Makefile:136: all] Error 2
[2025-07-23T18:57:31.713Z] gmake[4]: Leaving directory '/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/vm'
[2025-07-23T18:57:31.713Z] gmake[3]: *** [/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/closed/OpenJ9.gmk:407: build-j9] Error 2
[2025-07-23T18:57:31.713Z] gmake[3]: Leaving directory '/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2'
[2025-07-23T18:57:31.713Z] gmake[2]: *** [/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/closed/custom/Main.gmk:49: j9vm-build] Error 2
[2025-07-23T18:57:31.713Z] gmake[2]: Leaving directory '/jenkins/workspace/Build_JDK21_s390x_zos_Personal@2'
[2025-07-23T18:57:31.713Z] + exitcode=2
[2025-07-23T18:57:31.713Z] + printf '\nERROR: Build failed for targets '\''images test-image debug-image'\'' in configuration '\''zos-s390x-server-release'\'' (exit code 2) \n'
[2025-07-23T18:57:31.713Z] + /bin/tee -a /jenkins/workspace/Build_JDK21_s390x_zos_Personal@2/build/zos-s390x-server-release/build.log
[2025-07-23T18:57:31.713Z] 
```

Deigue commented:
I see the file was edited yesterday https://github.com/eclipse-openj9/openj9/blob/master/runtime/compiler/optimizer/J9EstimateCodeSize.cpp ? Perhaps possible runtimes fork isn't up to date with the changes for the Personal build to work

keithc-ca commented:
Please spell "z/OS" consistently in the commit message and here.

babsingh reviewed (approved):
The Linux riscv64 failure is infra-related. Will wait on @keithc-ca's approval before merging.

babsingh reviewed (commented):
Will also wait on @Deigue's latest JDK21 build results.

keithc-ca commented:
> Please spell "z/OS" consistently in the commit message and here.

This has not been addressed.

babsingh commented:
@Deigue confirmed that JDK21 testing was successful with the latest changes. Restarting the PR builds.

jenkins build all
